cmd/oras/discover.go

```diff
@@ -0,0 +1,179 @@
+	"strings"
+
+	"oras.land/oras-go/v2/registry/remote"
+	"oras.land/oras/cmd/oras/internal/option"
+
+	"github.com/need-being/go-tree"
+	digest "github.com/opencontainers/go-digest"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	artifactspec "github.com/oras-project/artifacts-spec/specs-go/v1"
+	"github.com/spf13/cobra"
+)
+
+type discoverOptions struct {
+	option.Common
+	option.Remote
+
+	targetRef    string
+	artifactType string
+	outputType   string
+}
+
+func discoverCmd() *cobra.Command {
+	var opts discoverOptions
+	cmd := &cobra.Command{
+		Use:   "discover [options] <name:tag|name@digest>",
+		Short: "discover artifacts from remote registry",
+		Long: `discover artifacts from remote registry
+
+Example - Discover all the artifacts linked with the specified reference:
+  oras discover localhost:5000/hello
+
+Example - Discover all the artifacts linked with the specified reference in a tree view:
+  oras discover localhost:5000/hello -o tree
+
+Example - Discover artifacts of type test-artifact test-artifact linked with the specified reference:
+  oras discover --artifact test-artifact localhost:5000/hello
+`,
+		Args: cobra.ExactArgs(1),
+		PreRun: func(cmd *cobra.Command, args []string) {
+			fmt.Println("Command discover is experimental and under development.")
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			opts.targetRef = args[0]
+			return runDiscover(opts)
+		},
+	}
+
+	cmd.Flags().StringVarP(&opts.artifactType, "artifact-type", "", "", "artifact type")
+	cmd.Flags().StringVarP(&opts.outputType, "output", "o", "table", fmt.Sprintf("Format in which to display references (%s, %s, or %s). tree format will show all references including nested", "table", "json", "tree"))
+	option.ApplyFlags(&opts, cmd.Flags())
+	return cmd
+}
+
+func runDiscover(opts discoverOptions) error {
+	ctx, _ := opts.SetLoggerLevel()
+	repo, err := opts.NewRepository(opts.targetRef, opts.Common)
+	if err != nil {
+		return err
+	}
+
+	// discover artifacts
+	root := tree.New(opts.targetRef)
+	desc, err := repo.Resolve(ctx, repo.Reference.Reference)
+	if err != nil {
+		return err
+	}
+	desc, refs, err := getAllReferences(ctx, repo, desc, opts.artifactType, root, opts.outputType == "tree")
+	if err != nil {
+		return err
+	}
+
+	switch opts.outputType {
+	case "tree":
+		tree.Print(root)
+	case "json":
+		printDiscoveredReferencesJSON(desc, *refs)
+	default:
+		fmt.Println("Discovered", len(*refs), "artifacts referencing", opts.targetRef)
+		fmt.Println("Digest:", desc.Digest)
+
+		if len(*refs) != 0 {
+			fmt.Println()
+			printDiscoveredReferencesTable(*refs, opts.Verbose)
+		}
+	}
+	return nil
+}
+
+func getAllReferences(ctx context.Context, repo *remote.Repository, desc ocispec.Descriptor, artifactType string, node *tree.Node, queryGraph bool) (ocispec.Descriptor, *[]artifactspec.Descriptor, error) {
+	var results []artifactspec.Descriptor
+	err := repo.Referrers(ctx, desc, func(referrers []artifactspec.Descriptor) error {
```

Comment:
Let's upgrade `oras-go` version and use the latest referrer implementation: https://github.com/oras-project/oras-go/pull/211